src/parse/asp/exec.go

```diff
@@ -159,51 +151,48 @@ func execFindCmd(cmdName string) (path string, err error) {
 }
 
 // execGetCachedOutput returns the output if found, sets found to true if found,
-// and returns a held promise that must be either cancelled or completed.
-func execGetCachedOutput(key execKey, args []string) (output string, found bool) {
-	execCacheLock.RLock()
-	out, found := execCachedOuts[key]
-	execCacheLock.RUnlock()
+// and returns a held promise that must be completed.
+func execGetCachedOutput(key execKey, args []string) (output *execOut, found bool) {
+	outputRaw, found := execCachedOuts.Load(key)
 	if found {
-		return out, true
+		return outputRaw.(*execOut), true
 	}
 
-	// Re-check with exclusive lock held
-	execCacheLock.Lock()
-	out, found = execCachedOuts[key]
+	// Re-check with promises exclusive lock held
+	execPromisesLock.Lock()
+	outputRaw, found = execCachedOuts.Load(key)
 	if found {
-		execCacheLock.Unlock()
-		return out, true
+		return outputRaw.(*execOut), true
```

Comment:
*sigh*  This was just a mis-`git add -p`.  :-/

```
# Manual hunk edit mode -- see bottom for a quick guide.
@@ -162,6 +175,8 @@ func execGetCachedOutput(key execKey, args []string) (output *execOut, found boo
        execPromisesLock.Lock()
        outputRaw, found = execCachedOuts.Load(key)
        if found {
+               execPromisesLock.Unlock()
+               // q.Q(goid.Get(), "cache hit2", key, args, outputRaw.(execOut))
                return outputRaw.(*execOut), true
        }
```

Reply:
Anyway, to answer your question: I would certainly invite suggestions here.  I think this is just a straight-up bug in that I didn't commit the unlock.  The reason this isn't broken up further is because it actually makes the code significantly more difficult to read.  For example, here's something I tried locally and found the call-site much more difficult to read vs manually (and correctly) unlocking the mutex:

```go
// execGetCachedOutputLocked is a small helper function that bounds the lock
// domain of execPromisesLock.
func execGetCachedOutputLocked(key execKey, args []string) (out *execOut, promise *execPromise, found bool) {
```